compiler+runtime/src/cpp/jank/runtime/module/loader.cpp

```diff
@@ -338,47 +339,101 @@ namespace jank::runtime::module
 
-    string_result<void> res{ err(fmt::format("no sources for registered module: {}", module)) };
+    auto const with_module_type{ [&](module_type const type) {
+      loader::find_result res{ entry->second, type };
+      return res;
+    } };
 
     //fmt::println("loading nested module {}", module);
 
-    /* When we're compiling, we always load from source. Otherwise, we
-     * load from source if we specifically chose to do so. */
-    bool const compiling{ truthy(rt_ctx.compile_files_var->deref()) };
-    if(compiling || ori == origin::source)
+    if(ori == origin::source)
     {
       if(entry->second.jank.is_some())
       {
-        res = load_jank(entry->second.jank.unwrap());
+        return with_module_type(module_type::jank);
       }
       else if(entry->second.cljc.is_some())
       {
-        res = load_cljc(entry->second.cljc.unwrap());
+        return with_module_type(module_type::cljc);
       }
     }
     else
     {
-      /* TODO: origin must be latest, so we need to find that. Even if there
-       * is a binary, we need to only choose it if it's as new, or newer, than
-       * the source. */
-      if(entry->second.o.is_some())
+      if(entry->second.o.is_some()
+         && (entry->second.jank.is_some() || entry->second.cljc.is_some()))
       {
-        res = load_o(module, entry->second.o.unwrap());
+        auto const o_file_path{ native_transient_string{ entry->second.o.unwrap().path } };
+
+        native_transient_string source_path{};
+        module_type module_type{};
+
+        if(entry->second.jank.is_some())
+        {
+          source_path = entry->second.jank.unwrap().path;
+          module_type = module_type::jank;
+        }
+        else
+        {
+          source_path = entry->second.cljc.unwrap().path;
+          module_type = module_type::cljc;
+        }
+
+        if(boost::filesystem::last_write_time(o_file_path)
+           >= boost::filesystem::last_write_time(source_path))
+        {
+          return with_module_type(module_type::o);
+        }
+        else
+        {
+          return with_module_type(module_type);
+        }
       }
       else if(entry->second.cpp.is_some())
       {
-        res = load_cpp(entry->second.cpp.unwrap());
+        return with_module_type(module_type::cpp);
       }
       else if(entry->second.jank.is_some())
       {
-        res = load_jank(entry->second.jank.unwrap());
+        return with_module_type(module_type::jank);
       }
       else if(entry->second.cljc.is_some())
       {
-        res = load_cljc(entry->second.cljc.unwrap());
+        return with_module_type(module_type::cljc);
       }
     }
 
+    return err(fmt::format("no sources for registered module: {}", module));
+  }
+
+  string_result<void> loader::load(native_persistent_string_view const &module, origin const ori)
+  {
+    auto const &found_module{ loader::find(module, ori) };
+    if(found_module.is_err())
+    {
+      return err(std::move(found_module.expect_err()));
```

Comment:
There is `expect_err_move` for this.

Reply:
How will that work.

```
 'this' argument to member function 'expect_err_move' has type 'const string_result<loader::find_result>' (aka 'const result<jank::runtime::module::loader::find_result, native_persistent_string>'), but function is not marked const [member_function_call
 _bad_cvr]
```

Reply:
Right now, `found_module` is `const`, which means that you can't move from it. As the code is written now, you're calling `expect_err`, which returns an `E const  &` and then you're calling `std::move` on that, which does nothing since you can't move from an `E const &` since it's a ref to `const`.

Thus, the error using `expect_err_move` here is pointing out the actual error with you wanting to move here.

Overall, this is isn't a correct application of move, the way you have it. If you want it to be correct, use `expect_err_move`, but also, this is premature optimization in a case that my gut tells me is not worth the additional complexity.

Lastly, using `std::move` with return values is generally not a good idea unless those values can from parameters. If they're locals, you'll just get in the way of NRVO. If they're function call results, you'll either try to move something which can't be moved (like this code does) or get in the way of existing prvalues.

Reply:
TIL:
- Can't move consts
- std::move can disable NRVOs (Although I don't quite understand how NRVO works.)

I didn't understand this part:
> Lastly, using std::move with return values is generally not a good idea unless those values can from parameters

Reply:
No, that particular pattern is *not* a good idea, since `my_type &param1` doesn't convey that a move will happen. Instead, this would be better:

```cpp
my_type my_function(my_type &&param1)
{
   do_something_with_param1(param1);
   return std::move(param1);
}
```

For more info on why not to use `std::move` for returning non-params, see this:
https://stackoverflow.com/questions/19267408/why-does-stdmove-prevent-rvo-return-value-optimization